Volume rendering enhancements (Hackerthon results)

moreApi opened the pull request:
- enhanced transfer function GUI 
  - double click to delete or create point
  - dont move point past other point (if you are not going tooo fast)
- color map editor
- load from and safe to file for color map and tf+display range
- option for fixed distance ray cast stepping
- gpu histogram calculation for supported volumes 


The commit history is a bit screwed, because this is based on some code that got alreadz merged but the reference to those comit is lost, probatly because some rebasing or squashing.

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+     * Save a color map to file as a png.
+     */
+    fun saveToFile(file: File){
+        var file_ = file
```
Don't use underscores please, just call it `f` or something.

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+    /**
+     * A filter to only select pngs.
+     */
+    class PNGFileFilter: FileFilter()
```
This is exposed publicly, is that intended?

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+    /**
+     * Load a color map from png.
+     */
+    fun loadFromFile(file: File){
```
This should rather be a factory method that returns a Colormap

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+/**
+ * A Gui element to allow users to visually create or modify a color map
+ */
+class ColorMapEditor(var target:Volume? = null) : JPanel() {
```
For separation of concerns, this should be in its own file.

moreApi replied:
its now a sub class of color colorMapPanel.

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+}
+
+/**
+ * A Gui element to allow users to visually create or modify a color map
```
Typo, (GUI)

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+        target?.let { loadColormap(it.colormap) }
+
+        this.addMouseListener(object : MouseListener {
+            override fun mouseClicked(e: MouseEvent) {
```
Missing docs

moreApi replied:
I am a bit confused about the standart of documentation we are aming for. We are not enforcing documentation on private methods (which is the correct way to go IMO) then why are we enforcing documentation on an overwrite of an GUI Listener interface?

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+                repaint()
+            }
+
+            override fun mousePressed(e: MouseEvent) {
```
Missing docs

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+                dragged = false
+            }
+
+            override fun mouseReleased(e: MouseEvent?) {
```
Missing docs

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+                dragging = null
+            }
+
+            override fun mouseEntered(e: MouseEvent?) {/*noop*/}
```
Missing docs

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+            }
+
+            override fun mouseEntered(e: MouseEvent?) {/*noop*/}
+            override fun mouseExited(e: MouseEvent?) {/*noop*/}
```
Missing docs

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+        })
+
+        this.addMouseMotionListener(object : MouseMotionListener {
+            override fun mouseDragged(e: MouseEvent) {
```
Missing docs

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+                repaint()
+            }
+
+            override fun mouseMoved(e: MouseEvent) {
```
Missing docs

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/DisplayRangeEditor.kt:
```diff
@@ -0,0 +1,92 @@
+/**
+ * GUI for editing the display range of volumes.
+ * Part of TransferFunctionEditor
+ */
```
Author missing

aryaman-gupta replied:
0bacab27966c3598c202ad8f93d6a1e052de3e18 adds author info

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/DisplayRangeEditor.kt:
```diff
@@ -0,0 +1,92 @@
+            "fill",
+            "[left, 10%]5[right, 40%]5[left, 10%]5[right, 40%]"
+        )
+        val title = BorderFactory.createTitledBorder("Display Range")
```
Combine with assignment on the next line, also that one has an unnecessary `this`.

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/DisplayRangeEditor.kt:
```diff
@@ -0,0 +1,92 @@
+    private val maxValueLabel: JLabel
+
+    init {
+        this.layout = MigLayout(
```
Unnecessary `this`.

skalarproduktraum replied:
In addition, MigLayout's constraints are not exactly intuitive. I just spend some time yesterday with them 😂 Could you maybe add a comment that clarifies what those are intended to achieve?

moreApi replied:
You are completly right. I also have no Idea what they do. I just copied them from here ;) https://github.com/scenerygraphics/scenery/blame/2fc22df0e998fc8c0f641cacd993a2e707f61fcc/src/main/kotlin/graphics/scenery/volumes/TransferFunctionEditor.kt#L344

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/HasTransferFunction.kt:
```diff
@@ -1,5 +1,10 @@
+import java.io.InputStream
+
 /**
  * @author Konrad Michel <Konrad.Michel@mailbox.tu-dresden.de>
```
I guess it has more authors now :-)

aryaman-gupta replied:
0bacab27966c3598c202ad8f93d6a1e052de3e18 adds author info

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/HasTransferFunction.kt:
```diff
@@ -11,4 +16,43 @@ interface HasTransferFunction {
+    /**
+     * Load transfer function and display range from file that was written by [HasTransferFunction.toFile]
+     */
+    fun loadTFFromFile(file: File){
```
Spell transfer function out, please

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/HasTransferFunction.kt:
```diff
@@ -11,4 +16,43 @@ interface HasTransferFunction {
+            val line = it.trim().split(";").mapNotNull(kotlin.String::toFloatOrNull)
+            if (line.size == 2){
+                if (!isRangeSet){
+                    this.minDisplayRange = line[0]
```
Unnecessary this

skalarproduktraum replied:
Likewise on the next line

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/HasTransferFunction.kt:
```diff
@@ -11,4 +16,43 @@ interface HasTransferFunction {
+                }
+            }
+        }
+        this.transferFunction = tf
```
Unnecessary this

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/HasTransferFunction.kt:
```diff
@@ -11,4 +16,43 @@ interface HasTransferFunction {
+     * First line is the display range sepaerated by a semicolon.
+     * All following lines are tf control points.
+     */
+    fun toFile(file: File){
```
This should rather be called `saveTransferFunctionToFile()`, otherwise this might indicate serialisation of the entire object

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/HasTransferFunction.kt:
```diff
@@ -11,4 +16,43 @@ interface HasTransferFunction {
+     */
+    fun toFile(file: File){
+        val writer = BufferedWriter(FileWriter(file))
+        writer.write("${this.minDisplayRange};${this.maxDisplayRange}")
```
Unnecessary `this`

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/HasTransferFunction.kt:
```diff
@@ -11,4 +16,43 @@ interface HasTransferFunction {
+        val writer = BufferedWriter(FileWriter(file))
+        writer.write("${this.minDisplayRange};${this.maxDisplayRange}")
+        writer.newLine()
+        this.transferFunction.controlPoints().forEach {
```
Unnecessary this

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/HistogramChartManager.kt:
```diff
@@ -0,0 +1,154 @@
+
+/**
+ * Handles all histogram related things.
+ */
```
Author missing

aryaman-gupta replied:
0bacab27966c3598c202ad8f93d6a1e052de3e18 adds author info

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/HistogramChartManager.kt:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    /**
+     * Old code from PowerOfNames. Calculates the histogram on the CPU. I distance myself from it.
```
That last sentence is completely unnecessary

skalarproduktraum replied:
Is this actually still necessary to have here? If not, remove it.

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/HistogramChartManager.kt:
```diff
@@ -0,0 +1,154 @@
+        volumeHistogramData.adjustForBinSize = false
+
+        if (tfContainer is HasHistogram) {
+            genHistButton.addActionListener() {
```
Great that this has become an ActionListener instead of a ChangeListener, I've just worked around that bug in sciview 👍

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/RaycastingPropertiesEditor.kt:
```diff
@@ -0,0 +1,65 @@
+
+    }
+
+    companion object{
```
Missing docs, and missing space before `{`

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/RaycastingPropertiesEditor.kt:
```diff
@@ -0,0 +1,65 @@
+        /**
+         * Convenience function to open a JFrame containing a [RaycastingPropertiesEditor]
+         */
+        fun showRaycastingProperties(volumeManager: VolumeManager){
```
Just call it `show()`

skalarproduktraum replied:
Also, as a factory method, this should return the editor component

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/TransferFunctionEditor.kt:
```diff
@@ -166,10 +123,10 @@ class TransferFunctionEditor constructor(
-            override fun mouseClicked(e: MouseEvent) {}
-            override fun mouseEntered(e: MouseEvent) {}
-            override fun mouseExited(e: MouseEvent) {}
+            override fun mousePressed(e: MouseEvent) {/*noop*/}
```
These guys should have docs on what they're (not) doing

moreApi replied:
like: "if a mouse button is pressed nothing happens because we dont need this functionality at the moment"? really?

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/TransferFunctionEditor.kt:
```diff
@@ -198,7 +154,7 @@ class TransferFunctionEditor constructor(
                         "%.2f / %.2f".format(mouseTargetCP.x.toFloat(), mouseTargetCP.y.toFloat()),
                         mouseTargetCP.x,
-                        mouseTargetCP.y)
+                        mouseTargetCP.y-0.04)
```
Where is the 0.04 coming from? Please add a comment why it's there

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/TransferFunctionEditor.kt:
```diff
@@ -211,10 +167,8 @@ class TransferFunctionEditor constructor(
                 }
             }
-            override fun mouseMoved(e: MouseEvent) {}
+            override fun mouseMoved(e: MouseEvent) {/*noop*/}
```
docs

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/Volume.kt:
```diff
@@ -209,6 +209,15 @@ open class Volume(
+            return when(dataSource){
+                VolumeDataSource.NullSource -> 1
+                is VolumeDataSource.RAISource<*> -> dataSource.type.toBytesPerValue()
+                is SpimDataMinimalSource -> (dataSource.sources.first().spimSource.type as NumericType<*>).toBytesPerValue() //I dont know if this works
```
If you don't know if this works, please test it. Let's not have some untested code here that might bite down the line 👍

moreApi replied:
I dont know how SpimDataMinimalSource work and it is not documented.

skalarproduktraum replied:
This is used every time you load a BDV format (XML/HDF5) dataset, you can therefore test it with any of these.

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/VolumeHistogramComputeNode.kt:
```diff
@@ -0,0 +1,153 @@
+import java.nio.ByteBuffer
+import java.nio.IntBuffer
+
+/**
```
Author(s) missing

aryaman-gupta replied:
0bacab27966c3598c202ad8f93d6a1e052de3e18 adds author info

skalarproduktraum commented on src/main/resources/graphics/scenery/volumes/ComputeHistogram.comp:
```diff
@@ -0,0 +1,47 @@
+void main() {
+
+    if(gl_GlobalInvocationID.xy == debug_pixel){
+        debugPrintfEXT("Hello World!!!!!!!!!!!!!!!!!!!!!!!!!!");
```
This is missing at least 10 exclamation marks ⚠️

No, but seriously, remove this :-D

aryaman-gupta replied:
115eb7066dd07530e6125e03f7002dab377d1f16 adds necessary exclamation marks

No, removes debug code :)

skalarproduktraum commented on src/main/resources/graphics/scenery/volumes/ComputeHistogram.comp:
```diff
@@ -0,0 +1,47 @@
+            int binID = int(floor(val / binSize));
+
+            int ret = imageAtomicAdd(Histogram, ivec3(binID, 0, 0), 1);
+            if( binID == 221 && ret != 0){
```
Remove debug code

aryaman-gupta replied:
115eb7066dd07530e6125e03f7002dab377d1f16 removes debug code

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/Colormap.kt:
```diff
@@ -155,7 +155,7 @@ class Colormap(val buffer: ByteBuffer, val width: Int, val height: Int) {
         @JvmStatic fun list(): List<String> {
             // FIXME: Hardcoded for the moment, not nice.
-            val list = arrayListOf("grays", "hot", "jet", "plasma", "viridis")
+            val list = arrayListOf("grays", "hot", "jet", "plasma", "viridis", "red-blue", "rb-darker")
```
Please see that the files you've added don't violate any licenses. Also, the PNG files contain some text and a border - both ought to be removed.

aryaman-gupta replied:
I got the colormaps from https://www.ncl.ucar.edu/Document/Graphics/color_table_gallery.shtml. Its an open-source scivis package and the license information is here https://www.ncl.ucar.edu/Download/NCL_source_license.shtml. My interpretation would be that we are allowed to use the files but need to include their license message. This is what they say about the code, not sure it also applies to the png files.

Do you suggest removing the files?

skalarproduktraum replied:
This is basically the BSD license, so it's fine, and we'll acknowledge them appropriately 👍

skalarproduktraum replied:
I have added an ACKNOWLEDGEMENTS.md file, and also removed the margin text on the colormaps.

skalarproduktraum reviewed (changes requested):
Needs some polish, but otherwise great work, thanks @moreApi and @aryaman-gupta 👍

skalarproduktraum commented:
I have now also played around with it a bit - and it's super nice! I do have however some minor complaints about the GUI:

<img width="478" alt="Screenshot 2024-02-21 at 11 00 57" src="https://github.com/scenerygraphics/scenery/assets/586495/9ecf6ed5-b59b-47aa-8fdc-bdf024309602">
* Scaling is broken once the window goes below a certain size - likely something has a minimum size set. This is a problem when embedding this in sciview's inspector.
* I am not a fan of the borders for Display Range and Color Map. I feel they're both self-explanatory and also don't need a label, or at least not one that wastes so much space.
* The colormap label before the drop down is not necessary.
* The load and save buttons should just be called load/save, and should be aligned to the right
* The control points for the color maps are often badly visible, could you change them to a filled dot, or a crosshair or something?

skalarproduktraum commented:
And one more: When saving a transfer function, the file does not have an extension by default. How about `.transferfunction`?

moreApi commented:
> * I am not a fan of the borders for Display Range and Color Map. I feel they're both self-explanatory and also don't need a label, or at least not one that wastes so much space.
I am a big fan of them. They make clear what belongs together in a way whitespace could never acheive. I like them so much that I think they are worth the space investment. But most elements are contained in their own Jpanel so for sciView the borders may be adjusted as desired.
> * The load and save buttons should just be called load/save , and should be aligned to the right
Since we have two different load/save buttons pairs very close to each other I prefere to have more speaking buttons. 
> * The control points for the color maps are often badly visible, could you change them to a filled dot, or a crosshair or something?
Since they are basically hand drawn this is not easily possible.

moreApi commented:
>Scaling is broken once the window goes below a certain size - likely something has a minimum size set. This is a problem when embedding this in sciview's inspector.

The problem are the buttons. But I am unable to change it so that the buttons stack, when there is not enoutgh space. Internet says MigLayout does not support this and I also did not got it to work with FlowLayout which should be perfect for it.

moreApi commented:
Also I dont agree with both codacy issues. plz ignore

aryaman-gupta commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,332 @@
+
+/**
+ * A JPanel to display everything related to setting and editing the color map of a volume.
+ */
```
0bacab27966c3598c202ad8f93d6a1e052de3e18 adds author info

aryaman-gupta commented:
I have added a commit 609d359c5f8c2cce7eb36e2a5f0c9824821ab010 that eliminates the final attenuation of color for volume rendering. I do not know where this comes from, but I don't see the point of it and it is, to my knowledge, not consistent with the mathematics of volume rendering.

skalarproduktraum commented on src/main/kotlin/graphics/scenery/volumes/ColorMapPanel.kt:
```diff
@@ -0,0 +1,313 @@
+    /**
+     * A GUI element to allow users to visually create or modify a color map
+     */
+    class ColorMapEditor(var target:Volume? = null) : JPanel() {
```
Consistency, as above

skalarproduktraum commented:
I made a few little changes to the UI:
![image](https://github.com/scenerygraphics/scenery/assets/586495/e3b5dc1b-2eab-4e26-98fb-8fd56c404210)

* control points for the colors are rendered as boxes, I feel they look nicer and render more consistently
* control points that are hovered over get filled
* the color of the control point's border is determined by the brightness of it's color, if it's dark, the border is white, otherwise it's black
* I've also modified the border a bit so it's basically a separator now

What do you think @moreApi and @aryaman-gupta?

skalarproduktraum commented:
What I'd like to do further is move the Load/Save buttons to the right, that also saves a bit of vertical space below the transfer function editor.

moreApi commented:
> * the color of the control point's border is determined by the brightness of it's color, if it's dark, the border is white, otherwise it's black
This is good.

Personal taste wise i prefere circles over boxes.

I am ok with the border alternative. 

Free floating save and load buttons look desperate for some context and need more explanation for example by adding what they save and load to their label, especially the first pair.

@skalarproduktraum 
(general question: should we tag each other? I check the PR anyway if it pops up in my githup notifications.)

skalarproduktraum commented:
I played around some more and I think I like this solution best:
![image](https://github.com/scenerygraphics/scenery/assets/586495/7c60585f-7349-4e99-aca0-28ea2b73ba45)

With regard to the tagging question, @moreApi, not in general, but as this was a specific question for you two, I decided to tag you ^^

skalarproduktraum commented:
So I think this is what I am halfway happy with now:
<img width="808" alt="image" src="https://github.com/scenerygraphics/scenery/assets/586495/89906a9a-7406-4e9b-8a02-c5e1e2a195e9">

* histogram domain axis values get scaled by the data range
* the histogram color is more subtle
* loading colormaps now works correctly

It seems though that loading colormaps does not work well, at least there's information missing when I load in one that I've created before.

skalarproduktraum commented:
Do @kephale or @ctrueden maybe have comments on the GUI?

kephale commented:
It looks great. 

One thing I was wondering about was adjusting the X-range of the histogram (posted here https://github.com/scenerygraphics/sciview/issues/558).

moreApi commented:
That is already the case. The x-axis is labeld wrong. Also one needs to refresh the histogram for a new scaling to take effect. refreshing is currently done by toggeling the histogram button

moreApi commented:
I fixed the old histogram code for spim sources and moved histogram generating code to their respective classes using the interface in a sensible way. Now this part should be more clean. 

Also I changed the color of the histogram to a more visible if not pretty color. We should prefere function over form. (we could think about keeping the background simple a white if we want to keep gray histogram bars.) The save/load buttons for transferfunctions got a litte "TF". Also not the most pretty solution but if avoids confusing the user with contextless buttons.

So far I am done again.

skalarproduktraum commented:
Nice, thanks @moreApi! Next time, please post a screenshot of the current state as well.

I gotta say I am not a big fan of using "TF" as an abbreviation. What I have done now is implemented the load/save functionality we have under gear menu:

![image](https://github.com/scenerygraphics/scenery/assets/586495/c0053f8e-0636-49be-a11d-a7e73744d3dd)

I think this is quite a nice solution, and also opened the possibility to reset the transfer function, because there's more space now. What do you think?

moreApi commented:
I was also not a fan of the "TF" but it was better than no context. BUT that cogwheel is awsome and solves all of our problems including space 👍🎆

kephale commented:
> I gotta say I am not a big fan of using "TF" as an abbreviation.

WTF? What TF?
